util/src/main/kotlin/com/kamelia/sprinkler/util/Box.kt

```diff
@@ -0,0 +1,154 @@
+package com.kamelia.sprinkler.util
+
+import com.zwendo.restrikt.annotation.HideFromJava
+
+/**
+ * Represents a box containing a value. This interface aims at providing a way to declare a property with a value that
+ * is not available at the time of declaration. Any read access to the [value] property will throw an exception if the
+ * value has not been set.
+ *
+ * It differs from [Lazy] in that the value can be set from outside the class. A typical use case is to declare an
+ * object containing a property subject to a circular dependency without having to resort to a mutable property in the
+ * class.
+ *
+ * Example:
+ * ```
+ * class Foo(intBox: Box<Int>) {
+ *
+ *     val i by intBox
+ *
+ * }
+ *
+ * fun main() {
+ *    val box = Box.SingleWriteBox<Int>()
+ *    val foo = Foo(box)
+ *
+ *    runCatching {
+ *      println(foo.i) // Throws an exception
+ *    }
+ *
+ *    box.fill(1)
+ *    println(foo.i) // Prints 1
+ *
+ * }
+ *
+ * ```
+ *
+ * @param T the type of the value
+ * @see Box.Mutable
+ * @see Box.SingleWriteBox
+ * @see Box.RewritableBox
+ * @see Box.PrefilledBox
+ */
+interface Box<T> {
```

Comment:
could make factories for each box type, in the same vein as listOf, setOf, etc

that's also good for java, as it will avoid that dumdumb empty diamond